File: drone_nav/drone_nav_path_planner/src/local_planner.cpp
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <geometry_msgs/msg/point.hpp>
#include <nav_msgs/msg/path.hpp>
#include "drone_nav_path_planner/srv/pos.hpp"
#include <visualization_msgs/msg/marker_array.hpp>
#include <octomap/OcTree.h>
#include <octomap/octomap.h>

#include <mutex>

using namespace std::chrono_literals;

class localplanner : public rclcpp::Node     
{
public:
  localplanner() : Node("local_planner_server")
  {
    this->declare_parameter<double>("voxel_resolution", 0.2);
    this->declare_parameter<double>("path_footprint", 0.5);

    double voxel_resolution = this->get_parameter("voxel_resolution").as_double();
    path_footprint= this->get_parameter("path_footprint").as_double();
    
    RCLCPP_INFO(this->get_logger(),"%.2f",path_footprint);

    local_voxel_sub = this->create_subscription<visualization_msgs::msg::MarkerArray>(
        "/local_occupancy_grid", 10, std::bind(&localplanner::local_voxel_callback, this, std::placeholders::_1));

    trajectory_sub = this->create_subscription<nav_msgs::msg::Path>(
        "/nav_trajectory", 10, std::bind(&localplanner::trajectory_callback, this, std::placeholders::_1));

    planner_client = this->create_client<drone_nav_path_planner::srv::Pos>("trajectory_planner");

    while (!planner_client->wait_for_service(1s)) {
      RCLCPP_INFO(this->get_logger(), "Waiting for service...");
    }
  }

private:
  bool isCollision(float x, float y, float z) {

    resolution = temp_tree.getResolution();

    float step = resolution / 2.0;
    for (float dx = -path_footprint; dx <= path_footprint; dx += step) {
        for (float dy = -path_footprint; dy <= path_footprint; dy += step) {
            for (float dz = -path_footprint; dz <= path_footprint; dz += step) {
                if (std::sqrt(dx*dx + dy*dy + dz*dz) > path_footprint) continue;
                octomap::point3d p(x + dx, y + dy, z + dz);
                octomap::OcTreeNode* node1 = temp_tree.search(p);
                if (node1 == nullptr) {
                  continue;
                }
                bool occupied = node1 != nullptr && temp_tree.isNodeOccupied(node1);
              
                if (occupied) return true;
            }
        }
    }
    return false; 
  }

  void trajectory_callback(const nav_msgs::msg::Path::SharedPtr msg) {
    std::lock_guard<std::mutex> lock(tree_mutex_);

      for (const auto& pose_stamped : msg->poses) {
        double x = pose_stamped.pose.position.x;
        double y = pose_stamped.pose.position.y;
        double z = pose_stamped.pose.position.z;
        if (temp_tree.size() == 0) {
          return;
        }
        if (isCollision(x, y, z)){
          RCLCPP_WARN(this->get_logger(),"CAN COLLIDE");
          std::cout<<"can collide" << std::endl;

          auto request = std::make_shared<drone_nav_path_planner::srv::Pos::Request>();
          request->state = true;
          request->goal_position.x = 10.0;
          request->goal_position.y = 20.9;
          request->goal_position.z = 2.0;

          planner_client->async_send_request(request,
            [this](rclcpp::Client<drone_nav_path_planner::srv::Pos>::SharedFuture result) {
              RCLCPP_INFO(this->get_logger(), "Received response from planner");
            });
          break;
      }
      }
    }
  // void trajectory_callback(const nav_msgs::msg::Path::SharedPtr msg) {
  //   std::lock_guard<std::mutex> lock(tree_mutex_);

  //     for (const auto& pose_stamped : msg->poses) {
  //       double x = pose_stamped.pose.position.x;
  //       double y = pose_stamped.pose.position.y;
  //       double z = pose_stamped.pose.position.z;
  //       if (temp_tree.size() == 0) {
  //         return;
  //       }
  //       octomap::OcTreeNode* node1 = temp_tree.search(x, y, z);
  //       if (node1 == nullptr) {
  //         continue;
  //       }
  //       if (temp_tree.isNodeOccupied(node1)) {

  //         std::cout<<"can collide" << std::endl;

  //         auto request = std::make_shared<drone_nav_path_planner::srv::Pos::Request>();
  //         request->state = true;
  //         request->goal_position.x = 20.1;
  //         request->goal_position.y = 8.9;
  //         request->goal_position.z = 2.0;

  //         planner_client->async_send_request(request,
  //           [this](rclcpp::Client<drone_nav_path_planner::srv::Pos>::SharedFuture result) {
  //             RCLCPP_INFO(this->get_logger(), "Received response from planner");
  //           });
  //         break;
  //     }
  //     }
  //   }
  

  void local_voxel_callback(const visualization_msgs::msg::MarkerArray::SharedPtr msg) {
    std::lock_guard<std::mutex> lock(tree_mutex_);
    temp_tree.clear();  

    for (const auto& marker : msg->markers) {
      for (const auto& pt : marker.points) {
        if (!std::isnan(pt.x) && !std::isnan(pt.y) && !std::isnan(pt.z)) {
          octomap::point3d p(pt.x, pt.y, pt.z);
          temp_tree.updateNode(p, true);  
        }
      }
    }
  }

  rclcpp::Client<drone_nav_path_planner::srv::Pos>::SharedPtr planner_client;
  rclcpp::Subscription<nav_msgs::msg::Path>::SharedPtr trajectory_sub;
  rclcpp::Subscription<visualization_msgs::msg::MarkerArray>::SharedPtr local_voxel_sub;
  float resolution;
  octomap::OcTree temp_tree{0.2};
  std::mutex tree_mutex_;
  double path_footprint;
};

int main(int argc, char *argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<localplanner>());
  rclcpp::shutdown();
  return 0;
}

File: drone_nav/drone_nav_path_planner/temp/local_planner.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from nav_msgs.msg import Path
from visualization_msgs.msg import MarkerArray
from geometry_msgs.msg import PoseStamped
import sys
import octomap
import numpy as np
from drone_nav_path_planner.srv import Pos
import math

def points_are_close(p1, p2, tol=1e-3):
    return all(math.isclose(a, b, abs_tol=tol) for a, b in zip(p1, p2))

class localplanner(Node):
    def __init__(self):
        super().__init__('local_planner')
        
        self.goal_position = []
        self.orientation = []
        self.goal_pos = [20.1, 8.9, 8.9]
        self.tree_resolution = 0.2

        self.trajectory_sub = self.create_subscription(Path, "/nav_trajectory",self.trajectory_callback, 10)
        self.local_obstacle_sub = self.create_subscription(MarkerArray, "/local_occupancy_grid", self.voxel, 10)

        # self.goal_pose = self.create_subscription(PoseStamped,"/goal_pose",self.position, 10)
        self.planner_client = self.create_client(Pos,"/trajectory_planner")


        while not self.planner_client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('planner service not available, waiting again...')

    # def position(self, msg):
    #     # print(msg.pose.position)
    #     # print(msg.pose.orientation)

    #     self.goal_position = [msg.pose.position.x, msg.pose.position.y, msg.pose.position.z]
    #     self.orientation = [msg.pose.orientation.x, msg.pose.orientation.y, msg.pose.orientation.z, msg.pose.orientation.w]


    def trajectory_callback(self, msg):
        self.trajectory_positions = []
        for pose_stamped in msg.poses:
            pt = pose_stamped.pose.position
            qt = pose_stamped.pose.orientation

            self.trajectory_positions.append((pt.x, pt.y, pt.z))
            # self.trajectory_orientations.append((qt.x, qt.y, qt.z, qt.w))
        

    def voxel(self, msg):
        try:
            self.obstacle = []
            self.tree = octomap.OcTree(self.tree_resolution)

            for marker in msg.markers:
                for point in marker.points:
                    # print(f"Point: {point}")

                    self.obstacle.append((point.x, point.y, point.z))
                    
            self.tree.insertPointCloud(
                np.array(self.obstacle), np.array((0.0, 0.0, 0.0))
            )
            for path in self.trajectory_positions:
                for node in self.tree.begin_leafs():
                    if self.tree.isNodeOccupied(node):

                        x, y, z = self.tree.keyToCoord(node.getKey())
                        voxel = [x, y ,z]
                        if points_are_close(path, voxel):
                            print(f"⚠ Obstacle near path point: {path} matches voxel {voxel}")
                        else:
                            print(f"Ea")
                        # break
                        # print(f"Occupied voxel at: x={x}, y={y}, z={z}")

            # for path in self.trajectory_positions:
            #     print(path)

            #     node = self.tree.search(path, 0)
            #     # print(node)
            #     # if node is None:
            #     #     continue
            #     try:
            #         if self.tree.isNodeOccupied(node):
            #             print("obstacle in the path")
            #             self.planner_service()
            #     except Exception as e:
            #         print(e)
            # #         # self.planner_service()
            # #         # break
            # #     else:
            # #         print("path is free")
            # # print(len(self.obstacle))
        except Exception as e:
            print(e)


    def planner_service(self):
        '''
            Description:    payload_service function is used to perform payload operation.
        '''
        if Pos is None:
            return
        req = Pos.Request()
        req.state = True
        req.goal_position.x = self.goal_pos[0]
        req.goal_position.y = self.goal_pos[1]
        req.goal_position.y = self.goal_pos[2]

        self.planner_client.call_async(req)

def main():
    rclpy.init(args=sys.argv)

    node = localplanner() 

    rclpy.spin(node)

    node.destroy_node()

    rclpy.shutdown()


if __name__ == '__main__':
    main()



